### PR TITLE
o/devicestate: copy seed on finish install API call, for UC

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Copy the seed on Ubuntu Core installations.

### CODE REVIEW

File: overlord/devicestate/handlers_install.go [hunk 1]
```diff
@@ -892,8 +892,36 @@ func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (
...
+	if output, stderr, err := osutil.RunSplitOutput("cp", "-rT", src, dst); err != nil {
+		return osutil.OutputErrCombine(output, stderr, err)
+	}
+	return nil
+}
+
+func copyLabelToSeedPartition(label, seedMntDir string) error {
+	// TODO we are copying all snaps from snaps/ folder, we should look at
+	// the assertions from the label to identify which are really needed,
+	// as more than one label might exist.
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:911]: it might sense to make installing/copying a seed a method of seed.Seed for the UC20+ variant exposed as other things via an interface

File: overlord/devicestate/handlers_install.go [hunk 2]
```diff
@@ -1010,6 +1012,30 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
...
 
+	if !sys.Model.Classic() {
+		sysSeed, err := seedOpen(dirs.SnapSeedDir, sys.Label)
+		if err != nil {
+			return fmt.Errorf("cannot open seed: %w", err)
+		}
+
+		if copier, ok := sysSeed.(seed.Copier); ok {
+			db := assertstate.DB(st)
+			err = copier.LoadAssertions(db, func(b *asserts.Batch) error {
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:1023]: we don't want to get these assertions into the main db I think, it should be fine for this seed to use a temporary db

File: overlord/devicestate/handlers_install.go [hunk 3]
```diff
@@ -1010,6 +1012,30 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
 	}
 	defer unmountParts()
 
+	if !sys.Model.Classic() {
+		sysSeed, err := seedOpen(dirs.SnapSeedDir, sys.Label)
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:1016]: double-checking, we don't open a seed elsewhere in this function?

File: overlord/devicestate/handlers_install.go [hunk 4]
```diff
@@ -1010,6 +1012,30 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
...
+			db := assertstate.DB(st)
+			err = copier.LoadAssertions(db, func(b *asserts.Batch) error {
+				return assertstate.AddBatch(st, b, nil)
+			})
+			if err != nil {
+				return fmt.Errorf("cannot load seed assertions: %w", err)
+			}
+
+			logger.Debugf("copying label %q to seed partition", sys.Label)
+			if err := copier.Copy(seedMntDir, sys.Label, timings.New(nil)); err != nil {
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:1031]: we have infrastructure to measure timing for tasks, maybe it's time to hook that up here

File: overlord/devicestate/handlers_install.go [hunk 5]
```diff
@@ -1010,6 +1012,30 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
...
+				return fmt.Errorf("cannot copy seed: %w", err)
+			}
+		} else {
+			logger.Debugf("seed does not support copying: %s", sys.Label)
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:1035]: that sound more like it should be an internal error, we expect UC20+ seeds to be copyable

**Review by pedronis — COMMENTED** [review]
thanks, some comments

File: overlord/devicestate/handlers_install.go (hunk 3 above)
Comment (bboozzoo) [reply overlord/devicestate/handlers_install.go:1016]: we do, but AFAITC it's not directly returned from `loadAndMountSystemLabelSnaps()`, perhaps a little refactor would do it

File: overlord/devicestate/handlers_install.go (hunk 2 above)
Comment (bboozzoo) [reply overlord/devicestate/handlers_install.go:1023]: Seed interface already specifies that passing nil as db will result in using a temporary DB, simply `copier.LoadAssertions(nil, nil)` should do it AFAIU

File: overlord/devicestate/handlers_install.go (hunk 3 above)
Comment (andrewphelpsj) [reply overlord/devicestate/handlers_install.go]: Moved things around so we only open it once now.

File: overlord/devicestate/devicemgr.go [hunk 6]
```diff
@@ -2130,21 +2130,21 @@ func (m *DeviceManager) SystemAndGadgetAndEncryptionInfo(wantedSystemLabel strin
 // includes system, gadget information, gadget and kernel snaps info,
 // and gadget and kernel seed snap info.
 func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type) (
-	*System, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, error) {
+	*System, seed.Seed, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, error) {
```
Comment (bboozzoo) [inline overlord/devicestate/devicemgr.go:2133]: hm this has become a bit hard to follow, could we try something like:
```
type systemAndEssentialSnaps struct {
    *System
    Seed seed.Seed,
    InfosByType map[snap.Type]*snapInfo
    SeedSnapsByType map[snap.Type]*seed.Snap
}
```

then this becomes:    
```suggestion
	*systemAndEssentialSnaps, error) {
```

File: overlord/devicestate/handlers_install.go [hunk 7]
```diff
@@ -854,12 +854,12 @@ func mountSeedSnap(seedSn *seed.Snap) (mountpoint string, unmount func() error,
 }
 
 func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (
-	*System, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, map[snap.Type]string, func(), error) {
+	*System, seed.Seed, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, map[snap.Type]string, func(), error) {
```
Comment (bboozzoo) [inline overlord/devicestate/handlers_install.go:857]: then we could give it the same treatment or just reduce to 3 returned values, your call

```suggestion
	sysInfo *systemAndEssentialSnaps, mountPoints map[snap.Type]string, func(), err error) {
```

File: overlord/devicestate/devicemgr.go (hunk 6 above)
Comment (andrewphelpsj) [reply overlord/devicestate/devicemgr.go:2133]: Good idea, I'll do this.

File: overlord/devicestate/handlers_install.go (hunk 7 above)
Comment (andrewphelpsj) [reply overlord/devicestate/handlers_install.go:857]: I think here 3 is probably not too bad.

File: overlord/devicestate/devicemgr.go (hunk 6 above)
Comment (andrewphelpsj) [reply overlord/devicestate/devicemgr.go]: Done!

**Review by bboozzoo — APPROVED** [review]
LGTM with a little tweak

File: overlord/devicestate/handlers_install.go [hunk 8]
```diff
@@ -853,17 +853,16 @@ func mountSeedSnap(seedSn *seed.Snap) (mountpoint string, unmount func() error,
 		nil
 }
 
-func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (
-	*System, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, map[snap.Type]string, func(), error) {
+func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (systemAndEssentialSnaps, map[snap.Type]string, func(), error) {
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/devicestate/handlers_install.go:856]: I generally try to reserve returing pointers for data types that either shouldn't be copied, are expensive to copy, or implement an interface. I'm happy to change it though, if you think returning the default structs makes it much worse to read.

[the remark this answers is not in the document]
Comment (Meulengracht) [reply overlord/devicestate/handlers_install.go:856]: hmm depends on the callers of this function, if it makes more sense to return an empty struct, then fine, but if the struct isn't actually used, then you keep making useless copies of an empty structure (and initialize it) whenever you return it. I think it makes more sense to return nil if the empty structure isn't needed.

**Review by Meulengracht — COMMENTED** [review]
+1, but still think the suggestion from @bboozzoo is valid

File: overlord/devicestate/handlers_install.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/devicestate/handlers_install.go]: I see, makes sense!

**Review by Meulengracht — APPROVED** [review]

File: overlord/devicestate/devicemgr.go [hunk 9]
```diff
@@ -2119,32 +2119,37 @@ func (m *DeviceManager) SystemAndGadgetAndEncryptionInfo(wantedSystemLabel strin
...
+	SeedSnapsByType map[snap.Type]*seed.Snap
 }
 
 // loadSystemAndEssentialSnaps loads information for the given label, which
 // includes system, gadget information, gadget and kernel snaps info,
 // and gadget and kernel seed snap info.
-func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type) (
-	*System, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, error) {
-
+func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type) (*systemAndEssentialSnaps, error) {
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:2139]: I wonder if this should take an option struct of some kind to control whether it returns the seed or not, I'm wondering this because seed is a relatively heavy object so memory wise dropping it earlier if not needed would be good?

**Review by pedronis — APPROVED** [review]
thanks, wondering about the lifetime/memory for the kept seed

File: overlord/devicestate/devicemgr.go (hunk 9 above)
Comment (andrewphelpsj) [reply overlord/devicestate/devicemgr.go:2139]: I'll address this in a followup PR.
